File: paleo_workbench/ui/components/facies_palette_widget.py
"""沉浸式相带编绘调色板（M2）——网格化相带画刷 + 装备状态呈现。

数据与颜色单一真源：词表 = :class:`FaciesTaxonomy`（工程覆盖优先），
颜色 = :func:`stage_actions.facies_category_color`（与分类渲染器同款），
花纹 = ``facies_patterns``（有 SVG 映射的相才有）。装备状态写入宿主注入
的 :class:`~paleo_workbench.ui.workstation.facies_selector.FaciesBrushContext`
（本部件不私藏状态）。
"""
from __future__ import annotations

from typing import Any

from PySide6.QtCore import QSize, Qt, Signal
from PySide6.QtGui import QColor, QIcon, QPainter, QPixmap
from PySide6.QtWidgets import (
    QFrame,
    QGridLayout,
    QHBoxLayout,
    QLabel,
    QScrollArea,
    QSizePolicy,
    QToolButton,
    QVBoxLayout,
    QWidget,
)

from paleo_workbench.mapping.facies_patterns import (
    pattern_id_for_facies,
    pattern_path_for_facies,
)
from paleo_workbench.mapping.facies_taxonomy import FaciesTaxonomy

#: 常用相带快捷键位数（1-9 数字键装备；00-decisions D6/D11）。
FAVORITE_KEY_COUNT = 9


def facies_color(name: str) -> str:
    """相名 → 分类渲染器同款颜色（单一取色真源，D11 三处一致）。"""
    from paleo_workbench.ui.workstation.stage_actions import facies_category_color

    return facies_category_color(name)


def _swatch_icon(color_hex: str, pattern_name: str | None) -> QIcon:
    """纯色底 + （有 SVG 时）花纹角标的调色板缩略图。"""
    pixmap = QPixmap(18, 18)
    pixmap.fill(Qt.GlobalColor.transparent)
    painter = QPainter(pixmap)
    painter.fillRect(1, 1, 16, 16, QColor(str(color_hex)))
    painter.setPen(QColor("#26364d"))
    painter.drawRect(0, 0, 17, 17)
    painter.end()
    return QIcon(pixmap)


class FaciesPaletteWidget(QWidget):
    """相带画刷调色板：相分区 + 亚相格 + 吸色管开关 + 装备栏。"""

    eyedropper_toggled = Signal(bool)

    def __init__(self, parent: QWidget | None = None) -> None:
        super().__init__(parent)
        self.setObjectName("FaciesPalette")
        self._taxonomy: FaciesTaxonomy | None = None
        self._brush: Any = None
        self._swatches: list[dict] = []  # {button, selection, top, sub}
        self._favorites: list[dict] = []

        layout = QVBoxLayout(self)
        layout.setContentsMargins(6, 6, 6, 6)
        layout.setSpacing(6)

        top_row = QHBoxLayout()
        top_row.setSpacing(4)
        self.eyedropper_button = QToolButton(self)
        self.eyedropper_button.setObjectName("FaciesEyedropperButton")
        self.eyedropper_button.setText("💡 吸色管")
        self.eyedropper_button.setToolTip(
            "激活后在地图上点击已有相带，一键吸取其相带属性/颜色/花纹")
        self.eyedropper_button.setCheckable(True)
        self.eyedropper_button.toggled.connect(self.eyedropper_toggled.emit)
        self._summary_label = QLabel("—", self)
        self._summary_label.setObjectName("FaciesPaletteSummary")
        top_row.addWidget(self.eyedropper_button)
        top_row.addStretch(1)
        top_row.addWidget(self._summary_label)
        layout.addLayout(top_row)

        self._scroll = QScrollArea(self)
        self._scroll.setWidgetResizable(True)
        self._scroll.setFrameShape(QFrame.Shape.NoFrame)
        self._host = QWidget(self._scroll)
        self._host_layout = QVBoxLayout(self._host)
        self._host_layout.setContentsMargins(0, 0, 0, 0)
        self._host_layout.setSpacing(8)
        self._host_layout.addStretch(1)
        self._scroll.setWidget(self._host)
        layout.addWidget(self._scroll, 1)

        self.equip_label = QLabel("未装备 — 点击格装备画刷", self)
        self.equip_label.setObjectName("FaciesEquipLabel")
        layout.addWidget(self.equip_label)

    # -- 数据装配 -------------------------------------------------------------

    def set_taxonomy(self, taxonomy: FaciesTaxonomy) -> None:
        self._taxonomy = taxonomy
        self._rebuild()

    def set_brush(self, brush: Any) -> None:
        """注入装备上下文（FaciesBrushContext；本部件只读+equip 写入）。"""
        if self._brush is not None:
            try:
                self._brush.equipped_changed.disconnect(self._on_equipped)
            except (TypeError, RuntimeError):
                pass
        self._brush = brush
        if brush is not None:
            brush.equipped_changed.connect(self._on_equipped)
        self._on_equipped(brush.selection() if brush is not None else {})

    def _rebuild(self) -> None:
        while self._host_layout.count() > 1:
            item = self._host_layout.takeAt(0)
            widget = item.widget()
            if widget is not None:
                widget.deleteLater()
        self._swatches = []
        self._favorites = []
        if self._taxonomy is None:
            return
        for top in self._taxonomy.names("facies"):
            section = QFrame(self._host)
            section.setObjectName("FaciesPaletteSection")
            section.setProperty("faciesName", top)
            section_layout = QVBoxLayout(section)
            section_layout.setContentsMargins(4, 4, 4, 4)
            section_layout.setSpacing(4)
            header = QHBoxLayout()
            header_label = QLabel(top, section)
            header_label.setObjectName("FaciesPaletteSectionHeader")
            pattern = pattern_path_for_facies(top)
            if pattern is not None:
                header_icon = QLabel(section)
                header_icon.setPixmap(
                    QIcon(str(pattern)).pixmap(QSize(20, 20)))
                header.addWidget(header_icon)
            color_chip = QLabel(section)
            chip = QPixmap(12, 12)
            chip.fill(QColor(facies_color(top)))
            color_chip.setPixmap(chip)
            header.addWidget(color_chip)
            header.addWidget(header_label)
            header.addStretch(1)
            section_layout.addLayout(header)

            grid = QGridLayout()
            grid.setSpacing(3)
            column = 0
            for sub in self._taxonomy.names("sub_facies", (top,)):
                selection = {
                    "facies": top, "sub_facies": sub, "micro_facies": "",
                }
                button = QToolButton(section)
                button.setObjectName("FaciesSwatchButton")
                button.setCheckable(True)
                button.setText(sub)
                button.setToolTip(f"{top} / {sub}")
                button.setIcon(_swatch_icon(facies_color(top),
                                            pattern_id_for_facies(top)))
                button.setToolButtonStyle(
                    Qt.ToolButtonStyle.ToolButtonTextBesideIcon)
                button.setSizePolicy(
                    QSizePolicy.Policy.Expanding, QSizePolicy.Policy.Fixed)
                button.clicked.connect(
                    lambda _checked=False, sel=dict(selection):
                    self.equip_by_selection(sel))
                entry = {
                    "button": button, "selection": selection, "top": top,
                    "sub": sub,
                }
                self._swatches.append(entry)
                if len(self._favorites) < FAVORITE_KEY_COUNT:
                    self._favorites.append(entry)
                grid.addWidget(button, len(self._favorites) // 3,
                               column)
                column = (column + 1) % 3
            section_layout.addLayout(grid)
            self._host_layout.insertWidget(self._host_layout.count() - 1,
                                           section)
        self._refresh_key_hints()
        n_f, n_s, n_m = self._taxonomy.counts()
        self._summary_label.setText(f"相 {n_f} · 亚相 {n_s} · 微相 {n_m}")

    def _refresh_key_hints(self) -> None:
        for index, entry in enumerate(self._favorites, start=1):
            entry["button"].setText(f"{index} {entry['sub']}")

    # -- 装备 -----------------------------------------------------------------

    def favorites(self) -> list[dict]:
        return [
            {"selection": dict(entry["selection"]),
             "label": f"{entry['top']} / {entry['sub']}"}
            for entry in self._favorites
        ]

    def equip_by_selection(self, selection: dict) -> None:
        if self._brush is not None:
            self._brush.equip(selection)

    def equip_by_index(self, index: int) -> None:
        if 0 <= index < len(self._favorites):
            self.equip_by_selection(dict(self._favorites[index]["selection"]))

    def _on_equipped(self, selection: dict) -> None:
        armed = bool(selection.get("facies"))
        if armed:
            parts = [selection.get("facies", ""),
                     selection.get("sub_facies", ""),
                     selection.get("micro_facies", "")]
            label = " / ".join(p for p in parts if p)
            self.equip_label.setText(f"装备：{label}")
        else:
            self.equip_label.setText("未装备 — 点击格装备画刷")
        current = self._current_swatch_key(selection)
        for entry in self._swatches:
            checked = self._swatch_key(entry) == current if armed else False
            entry["button"].setChecked(checked)

    @staticmethod
    def _swatch_key(entry: dict) -> tuple[str, str]:
        return entry["top"], entry["sub"]

    @staticmethod
    def _current_swatch_key(selection: dict) -> tuple[str, str]:
        return (str(selection.get("facies") or ""),
                str(selection.get("sub_facies") or ""))

    # -- 计数（测试/视觉回归 ------------------------------------------------------

    def section_count(self) -> int:
        return sum(
            1
            for i in range(self._host_layout.count())
            if self._host_layout.itemAt(i).widget() is not None
            and self._host_layout.itemAt(i).widget().objectName()
            == "FaciesPaletteSection"
        )

    def swatch_count(self) -> int:
        return len(self._swatches)

    def set_eyedropper_active(self, active: bool) -> None:
        """外部（吸色管控制器）回同步按钮态（抑制再广播）。"""
        self.eyedropper_button.blockSignals(True)
        self.eyedropper_button.setChecked(active)
        self.eyedropper_button.blockSignals(False)

File: libs/ui_widgets/src/facies_palette_widget.cpp
#include "pwb/ui_widgets/facies_palette_widget.hpp"

#include "pwb/ui_widgets/core/facies_patterns.hpp"
#include "pwb/ui_widgets/ui_context.hpp"

#include <QCryptographicHash>
#include <QFrame>
#include <QGridLayout>
#include <QHBoxLayout>
#include <QPainter>
#include <QPixmap>
#include <QSizePolicy>

namespace pwb::ui_widgets {

// Compile-time source root (same idiom as platform_services); when the
// definition is absent the pattern dir simply does not exist -> honest
// no-pattern, matching the Python fallback.
#ifndef PWB_SOURCE_DIR
#define PWB_SOURCE_DIR ""
#endif

namespace {

// Pattern asset dir: repo-relative geo-viz-engine location (same idiom as
// the Python FACIES_PATTERN_DIR); missing dir -> honest no-pattern.
QString facies_pattern_dir() {
    return QStringLiteral(PWB_SOURCE_DIR) +
           QStringLiteral(
               "/geo-viz-engine/packages/geoviz_well_log/geoviz_well_log"
               "/assets/patterns/facies");
}

QString qt_md5_hex(const std::string& utf8) {
    return QString::fromLatin1(QCryptographicHash::hash(
                                   QByteArray::fromStdString(utf8),
                                   QCryptographicHash::Md5)
                                   .toHex());
}

// 纯色底 + 描边调色板缩略图（_swatch_icon parity; CANVAS_INK border）。
QIcon swatch_icon(const QString& color_hex) {
    QPixmap pixmap(18, 18);
    pixmap.fill(Qt::transparent);
    QPainter painter(&pixmap);
    painter.fillRect(1, 1, 16, 16, QColor(color_hex));
    painter.setPen(QColor(palette_token("CANVAS_INK")));
    painter.drawRect(0, 0, 17, 17);
    painter.end();
    return QIcon(pixmap);
}

QString trim_qs(const QString& s) { return s.trimmed(); }

}  // namespace

QString facies_color(const QString& name, const QString& feature_color) {
    return QString::fromStdString(core::facies_category_color(
        name.toStdString(), feature_color.toStdString(),
        [](const std::string& s) { return qt_md5_hex(s).toStdString(); }));
}

// ---------------------------------------------------------------------------
// FaciesBrushContext
// ---------------------------------------------------------------------------

FaciesBrushContext::FaciesBrushContext(QObject* parent) : QObject(parent) {}

QVariantMap FaciesBrushContext::selection() const {
    QVariantMap values = selection_;
    if (!values.contains("facies")) values.insert("facies", "");
    if (!values.contains("sub_facies")) values.insert("sub_facies", "");
    if (!values.contains("micro_facies")) values.insert("micro_facies", "");
    return values;
}

void FaciesBrushContext::equip(const QVariantMap& selection) {
    QVariantMap values = this->selection();
    auto field = [&selection](const char* key) {
        const QVariant v = selection.value(QString::fromLatin1(key));
        return v.isValid() && !v.isNull() ? v.toString() : QString();
    };
    values.insert("facies", field("facies"));
    values.insert("sub_facies", field("sub_facies"));
    values.insert("micro_facies", field("micro_facies"));
    values.remove("level");
    const std::map<std::string, std::string> sel = {
        {"facies", values.value("facies").toString().toStdString()},
        {"sub_facies", values.value("sub_facies").toString().toStdString()},
        {"micro_facies",
         values.value("micro_facies").toString().toStdString()},
    };
    values.insert("level", QString::fromStdString(
                               core::FaciesTaxonomy::selection_level(sel)));
    if (armed_ && values == selection_) return;
    selection_ = values;
    armed_ = true;
    emit equipped_changed(selection_);
}

void FaciesBrushContext::clear() {
    if (!armed_) return;
    selection_.clear();
    armed_ = false;
    emit equipped_changed(QVariantMap());
}

// ---------------------------------------------------------------------------
// FaciesPaletteWidget
// ---------------------------------------------------------------------------

FaciesPaletteWidget::FaciesPaletteWidget(QWidget* parent) : QWidget(parent) {
    setObjectName(QStringLiteral("FaciesPalette"));
    auto* layout = new QVBoxLayout(this);
    layout->setContentsMargins(6, 6, 6, 6);
    layout->setSpacing(6);

    auto* top_row = new QHBoxLayout();
    top_row->setSpacing(4);
    eyedropper_button = new QToolButton(this);
    eyedropper_button->setObjectName(QStringLiteral("FaciesEyedropperButton"));
    eyedropper_button->setText(QStringLiteral("吸色管"));
    eyedropper_button->setToolTip(QStringLiteral(
        "激活后在地图上点击已有相带，一键吸取其相带属性/颜色/花纹"));
    eyedropper_button->setCheckable(true);
    connect(eyedropper_button, &QToolButton::toggled, this,
            &FaciesPaletteWidget::eyedropper_toggled);
    summary_label_ = new QLabel(QStringLiteral("—"), this);
    summary_label_->setObjectName(QStringLiteral("FaciesPaletteSummary"));
    top_row->addWidget(eyedropper_button);
    top_row->addStretch(1);
    top_row->addWidget(summary_label_);
    layout->addLayout(top_row);

    scroll_ = new QScrollArea(this);
    scroll_->setWidgetResizable(true);
    scroll_->setFrameShape(QFrame::NoFrame);
    host_ = new QWidget(scroll_);
    host_layout_ = new QVBoxLayout(host_);
    host_layout_->setContentsMargins(0, 0, 0, 0);
    host_layout_->setSpacing(8);
    host_layout_->addStretch(1);
    scroll_->setWidget(host_);
    layout->addWidget(scroll_, 1);

    equip_label = new QLabel(QStringLiteral("未装备 — 点击格装备画刷"), this);
    equip_label->setObjectName(QStringLiteral("FaciesEquipLabel"));
    layout->addWidget(equip_label);
}

void FaciesPaletteWidget::set_taxonomy(
    const core::FaciesTaxonomy& taxonomy) {
    taxonomy_ = taxonomy;
    has_taxonomy_ = true;
    rebuild();
}

void FaciesPaletteWidget::set_brush(FaciesBrushContext* brush) {
    if (brush_ != nullptr) {
        disconnect(brush_, &FaciesBrushContext::equipped_changed, this,
                   nullptr);
    }
    brush_ = brush;
    if (brush_ != nullptr) {
        connect(brush_, &FaciesBrushContext::equipped_changed, this,
                &FaciesPaletteWidget::on_equipped);
    }
    on_equipped(brush_ != nullptr ? brush_->selection() : QVariantMap());
}

void FaciesPaletteWidget::rebuild() {
    while (host_layout_->count() > 1) {
        QLayoutItem* item = host_layout_->takeAt(0);
        if (item->widget() != nullptr) item->widget()->deleteLater();
        delete item;
    }
    swatches_.clear();
    favorites_.clear();
    if (!has_taxonomy_) return;

    for (const std::string& top_std : taxonomy_.names("facies")) {
        const QString top = QString::fromStdString(top_std);
        auto* section = new QFrame(host_);
        section->setObjectName(QStringLiteral("FaciesPaletteSection"));
        section->setProperty("faciesName", top);
        auto* section_layout = new QVBoxLayout(section);
        section_layout->setContentsMargins(4, 4, 4, 4);
        section_layout->setSpacing(4);
        auto* header = new QHBoxLayout();
        auto* header_label = new QLabel(top, section);
        header_label->setObjectName(
            QStringLiteral("FaciesPaletteSectionHeader"));
        const auto pattern = core::pattern_path_for_facies(
            top_std, facies_pattern_dir().toStdString());
        if (pattern) {
            auto* header_icon = new QLabel(section);
            header_icon->setPixmap(
                QIcon(QString::fromStdString(*pattern)).pixmap(QSize(20, 20)));
            header->addWidget(header_icon);
        }
        auto* color_chip = new QLabel(section);
        QPixmap chip(12, 12);
        chip.fill(QColor(facies_color(top)));
        color_chip->setPixmap(chip);
        header->addWidget(color_chip);
        header->addWidget(header_label);
        header->addStretch(1);
        section_layout->addLayout(header);

        auto* grid = new QGridLayout();
        grid->setSpacing(3);
        int column = 0;
        for (const std::string& sub_std :
             taxonomy_.names("sub_facies", {top_std})) {
            const QString sub = QString::fromStdString(sub_std);
            QVariantMap selection;
            selection.insert("facies", top);
            selection.insert("sub_facies", sub);
            selection.insert("micro_facies", "");
            auto* button = new QToolButton(section);
            button->setObjectName(QStringLiteral("FaciesSwatchButton"));
            button->setCheckable(true);
            button->setText(sub);
            button->setToolTip(top + " / " + sub);
            button->setIcon(swatch_icon(facies_color(top)));
            button->setToolButtonStyle(Qt::ToolButtonTextBesideIcon);
            button->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Fixed);
            connect(button, &QToolButton::clicked, this,
                    [this, selection]() { equip_by_selection(selection); });
            const int index = int(swatches_.size());
            swatches_.push_back(Swatch{button, selection, top, sub});
            // 冻结语义：行号 = favorites 全局长度 // 3（跨分区共享；首 9 格
            // 之后恒为 3 —— 保留 Python 原状，见 ledger P-QUIRK）。
            if (int(favorites_.size()) < kFavoriteKeyCount) {
                favorites_.push_back(index);
            }
            grid->addWidget(button, int(favorites_.size()) / 3, column);
            column = (column + 1) % 3;
        }
        section_layout->addLayout(grid);
        host_layout_->insertWidget(host_layout_->count() - 1, section);
    }
    refresh_key_hints();
    const auto [n_f, n_s, n_m] = taxonomy_.counts();
    summary_label_->setText(QStringLiteral("相 %1 · 亚相 %2 · 微相 %3")
                                .arg(n_f)
                                .arg(n_s)
                                .arg(n_m));
}

void FaciesPaletteWidget::refresh_key_hints() {
    int index = 1;
    for (const int fav : favorites_) {
        swatches_[fav].button->setText(
            QStringLiteral("%1 %2").arg(index).arg(swatches_[fav].sub));
        ++index;
    }
}

QList<QVariantMap> FaciesPaletteWidget::favorites() const {
    QList<QVariantMap> out;
    for (const int fav : favorites_) {
        const Swatch& entry = swatches_[fav];
        QVariantMap copy = entry.selection;
        copy.insert("label", entry.top + " / " + entry.sub);
        out.append(copy);
    }
    return out;
}

void FaciesPaletteWidget::equip_by_selection(const QVariantMap& selection) {
    if (brush_ != nullptr) brush_->equip(selection);
}

void FaciesPaletteWidget::equip_by_index(int index) {
    if (index >= 0 && index < int(favorites_.size())) {
        equip_by_selection(swatches_[favorites_[index]].selection);
    }
}

QPair<QString, QString> FaciesPaletteWidget::swatch_key(const Swatch& entry) {
    return {entry.top, entry.sub};
}

QPair<QString, QString> FaciesPaletteWidget::current_swatch_key(
    const QVariantMap& selection) {
    return {selection.value("facies").toString(),
            selection.value("sub_facies").toString()};
}

void FaciesPaletteWidget::on_equipped(const QVariantMap& selection) {
    const bool armed = !selection.value("facies").toString().isEmpty();
    if (armed) {
        QStringList parts;
        for (const char* key : {"facies", "sub_facies", "micro_facies"}) {
            const QString v = selection.value(QString::fromLatin1(key)).toString();
            if (!v.isEmpty()) parts.append(v);
        }
        equip_label->setText(QStringLiteral("装备：") + parts.join(" / "));
    } else {
        equip_label->setText(QStringLiteral("未装备 — 点击格装备画刷"));
    }
    const auto current = current_swatch_key(selection);
    for (const Swatch& entry : swatches_) {
        const bool checked = armed && swatch_key(entry) == current;
        entry.button->setChecked(checked);
    }
}

int FaciesPaletteWidget::section_count() const {
    int count = 0;
    for (int i = 0; i < host_layout_->count(); ++i) {
        QWidget* w = host_layout_->itemAt(i)->widget();
        if (w != nullptr && w->objectName() == "FaciesPaletteSection") ++count;
    }
    return count;
}

void FaciesPaletteWidget::set_eyedropper_active(bool active) {
    eyedropper_button->blockSignals(true);
    eyedropper_button->setChecked(active);
    eyedropper_button->blockSignals(false);
}

}  // namespace pwb::ui_widgets
